src/uct/ib/ud/accel/ud_mlx5.c

```diff
@@ -269,125 +356,30 @@ uct_ud_mlx5_ep_am_zcopy(uct_ep_h tl_ep, uint8_t id, const void *header,
                         unsigned header_length, const uct_iov_t *iov,
                         size_t iovcnt, unsigned flags, uct_completion_t *comp)
 {
-    uct_ud_mlx5_ep_t *ep = ucs_derived_of(tl_ep, uct_ud_mlx5_ep_t);
-    uct_ud_mlx5_iface_t *iface = ucs_derived_of(tl_ep->iface,
-                                                uct_ud_mlx5_iface_t);
-    size_t ctrl_av_size = uct_ud_mlx5_ep_ctrl_av_size(ep);
-    uct_ud_send_skb_t *skb;
-    struct mlx5_wqe_ctrl_seg *ctrl;
-    struct mlx5_wqe_inl_data_seg *inl;
-    uct_ud_neth_t *neth;
-    size_t inl_size, wqe_size;
-
-    UCT_CHECK_AM_ID(id);
-    UCT_CHECK_IOV_SIZE(iovcnt, uct_ud_mlx5_max_am_iov(),
-                       "uct_ud_mlx5_ep_am_zcopy");
     UCT_CHECK_LENGTH(sizeof(uct_ud_neth_t) + header_length, 0,
                      UCT_IB_MLX5_AM_ZCOPY_MAX_HDR(UCT_IB_MLX5_AV_FULL_SIZE),
                      "am_zcopy header");
-    UCT_UD_CHECK_ZCOPY_LENGTH(&iface->super, header_length,
-                              uct_iov_total_length(iov, iovcnt));
-
-    uct_ud_enter(&iface->super);
-
-    skb = uct_ud_ep_get_tx_skb(&iface->super, &ep->super);
-    if (!skb) {
-        uct_ud_leave(&iface->super);
-        return UCS_ERR_NO_RESOURCE;
-    }
-
-    ctrl = iface->tx.wq.curr;
-    inl  = uct_ib_mlx5_txwq_wrap_exact(&iface->tx.wq,
-                                       UCS_PTR_BYTE_OFFSET(ctrl, ctrl_av_size));
-    inl_size = header_length + sizeof(*neth);
-    inl->byte_count = htonl(inl_size | MLX5_INLINE_SEG);
-
-    neth = (void*)(inl + 1);
-    uct_ud_am_set_neth(neth, &ep->super, id);
-    /* force ACK_REQ because we want to call user completion ASAP */
-    neth->packet_type |= UCT_UD_PACKET_FLAG_ACK_REQ;
-
-    uct_ib_mlx5_inline_copy(neth + 1, header, header_length, &iface->tx.wq);
-
-    wqe_size  = ucs_align_up_pow2(ctrl_av_size + inl_size + sizeof(*inl),
-                                  UCT_IB_MLX5_WQE_SEG_SIZE);
-    wqe_size += uct_ib_mlx5_set_data_seg_iov(&iface->tx.wq,
-                                             UCS_PTR_BYTE_OFFSET(ctrl, wqe_size),
-                                             iov, iovcnt);
-    ucs_assert(wqe_size <= UCT_IB_MLX5_MAX_SEND_WQE_SIZE);
-
-    UCT_UD_EP_HOOK_CALL_TX(&ep->super, neth);
-    uct_ud_mlx5_post_send(iface, ep, 0, ctrl, wqe_size,
-                          UCT_IB_MAX_ZCOPY_LOG_SGE(&iface->super.super));
-
-    skb->len = sizeof(*neth) + header_length;
-    memcpy(skb->neth, neth, sizeof(*neth));
-    memcpy(skb->neth + 1, header, header_length);
-    uct_ud_am_set_zcopy_desc(skb, iov, iovcnt, comp);
-
-    uct_ud_iface_complete_tx_skb(&iface->super, &ep->super, skb);
-    UCT_TL_EP_STAT_OP(&ep->super.super, AM, ZCOPY, header_length +
-                      uct_iov_total_length(iov, iovcnt));
-    uct_ud_leave(&iface->super);
-    return UCS_INPROGRESS;
+    return uct_ud_mlx5_ep_inline_iov_post(tl_ep, id,
+                                          /* inl. header */  header, 0,
```

Comment:
imo, passing header twice looks confusing and error prone

Reply:
fixed - passing dummy